wiki/Writing-tests-with-good-behavioural-coverage.md

```diff
@@ -357,705 +401,903 @@ The different scenarios and the expected outcomes are,
 
-    // Verify that an invalid file format results in an appropriate exception
-    val exception = assertThrows<IllegalArgumentException> {
-        account.withdraw("user", "password", 200.0, file = invalidFile)
-    }
-    assertThat(exception.message).contains("Invalid file format")
+  restaurant.applyDiscount(isMember, code, quarter)
+  restaurant.placeOrder(listOf("Burger"))
+
+  assertThat((outputStream.toString().trim()).contains("Total bill after discount: 4.5")
+}
+```
+
+Laying out the logic and conditions directly within the test makes it independent of external functions or files, which makes the test easier to understand, maintain, and debug.
+
+Unlike production code, where duplication is often avoided, in test code, it’s sometimes better to duplicate code if it leads to clearer, more understandable tests. This ensures that the behavior being tested is directly represented in the test case.
+
+## 3. Importance of Descriptive Test Names
+
+Naming test functions descriptively helps in identifying the purpose and scope of each test. Use names that reflect the specific behavior being tested.
+
+Oppia Android follows a naming convention where the test names should read like a sentence, and be consistent with other nearby test names to facilitate easily coming up with new tests. Consider using a format similar to the following for naming test functions:
+
+```
+testAction_withOneCondition_withSecondCondition_hasExpectedOutcome
+```
+
+```kotlin
+@Test
+fun testPlaceOrder_withValidItems_orderPlacedSuccessfully() {
+  // Test Logic: Order should be placed with given items
 }
 
 @Test
-fun testWithdraw_withUnavailableFile_throwsException() {
-    val unavailableFile = File("nonExistentFile.pdf")
+fun testPlaceOrder_withEmptyItems_orderNotPlaced() {
+  // Test Logic: Handle empty order gracefully
+}
 
-    // Verify that attempting to use a non-existent file results in an exception
-    val exception = assertThrows<IllegalArgumentException> {
-        account.withdraw("user", "password", 200.0, file = unavailableFile)
-    }
-    assertThat(exception.message).contains("File not found")
+@Test
+fun testCalculateBill_withValidItems_correctBillCalculated() {
+  // Test Logic: Calculate correct bill for ordered items
 }
 
 @Test
-fun testWithdraw_withDefaultReceipt_noReceiptPrinted() {
-    val output = ByteArrayOutputStream()
-    System.setOut(PrintStream(output))
+fun testApplyDiscount_withMemberAndValidCode_discountApplied() {
+  // Test Logic: Apply 10% discount for valid code and membership
+}
 
-    account.withdraw("user", "password", 200.0)
+@Test
+fun testApplyDiscount_withNonMemberAndValidCode_noDiscountApplied() {
+  // Test Logic: No discount applied for non-member
+}
 
-    // Ensure that no receipt is printed when the default receipt flag is used
-    assertThat(output.toString().trim()).doesNotContain("Receipt:")
-    System.setOut(System.out)
+@Test
+fun testApplyDiscount_withMemberAndQuarter2_discountApplied() {
```

Comment:
Updated it to `testApplyDiscount_withMemberAndValidCode_inQuarter2_discountApplied`, would this meet the requirement or should it be split for each condition as `testApplyDiscount_withMember_validCode_inQuarter2_discountApplied`, `testApplyDiscount_withNonMember_validCode_noDiscountApplied` .

Reply:
I think the new one is sufficient for demonstrating the example. It could, as you pointed out, be further split up or refined, but we don't enforce that strict of a standard (I was hoping more for a clear example than a set of definitions/rules).